File: src/main/scala/hk/edu/polyu/datamining/pamap2/database/DatabaseHelper.java
package hk.edu.polyu.datamining.pamap2.database;

import com.rethinkdb.RethinkDB;
import com.rethinkdb.net.Connection;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;

import java.util.HashMap;
import java.util.concurrent.TimeoutException;

/**
 * Created by beenotung on 1/25/16.
 */
public class DatabaseHelper {
    public static final Connection conn;
    public static final RethinkDB r = RethinkDB.r;
    public static final String hostname;
    public static final int port;
    public static final String dbname;

    static {
        Connection conn1;
        Config config = ConfigFactory.parseResources("database.conf");
        hostname = config.getString("rethinkdb.host");
        port = config.getInt("rethinkdb.port");
        dbname = config.getString("rethinkdb.dbname");
        try {
            conn1 = r.connection()
                    .hostname(hostname)
                    .port(port)
                    .db(dbname)
                    .connect();
        } catch (TimeoutException e) {
            conn1 = null;
            e.printStackTrace();
            System.err.println("Failed to connect to database, exiting...");
            System.exit(1);
        }
        conn = conn1;
    }

    public static HashMap<String, Object> createDatabaseIfNotExist() {
        return createDatabaseIfNotExist(dbname);
    }

    public static HashMap<String, Object> createDatabaseIfNotExist(String dbname) {
        return r.dbList().contains(dbname).do_(exist -> r.branch(
                exist,
                new HashMap<String, Object>() {{
                    put("created", 0);
                }},
                r.dbCreate(dbname)
                )
        ).run(conn);
    }

    public static final String TABLE_STATUS = "status";

    public static boolean hasInit() {
        boolean result = r.dbList().contains(dbname).do_(dbExist -> r.branch(
                dbExist,
                r.db(dbname).tableList().contains(TABLE_STATUS).do_(tableExist -> r.branch(
                        tableExist,
                        true,
                        false
                        )
                ),
                false
        )).run(conn);
        return result;
    }
}

File: src/main/scala/hk/edu/polyu/datamining/pamap2/database/DatabaseHelper_.scala
package hk.edu.polyu.datamining.pamap2.database

/**
  * Created by beenotung on 1/26/16.
  */

import com.rethinkdb.RethinkDB.r
import hk.edu.polyu.datamining.pamap2.database.DatabaseHelper.{conn, dbname}
import hk.edu.polyu.datamining.pamap2.utils.Lang._

object DatabaseHelper_ {
  def hasInit: Boolean = {
    val tableName: String = Tables.Status.name
    val fieldName: String = Tables.Status.Field.actionStatus.toString
    r.dbList().contains(dbname).do_(reqlFunction1(dbExist =>
      r.branch(
        dbExist,
        r.db(dbname).tableList().contains(tableName).do_(reqlFunction1(tableExist => r.branch(
          tableExist,
          r.db(dbname).table(tableName).withFields(fieldName).count().ge(1),
          false
        ))),
        false
      ))
    ).run(conn)
  }

  def init(currentStatus: String, nextStatus: String): Unit = {
    val statusTableName: String = Tables.Status.name
    val statusFieldName: String = Tables.Status.Field.actionStatus.toString
    /* drop and create database */
    r.dbList().contains(dbname).do_(reqlFunction1(dbExist => r.branch(
      dbExist,
      r.dbDrop(dbname),
      r.hashMap("dbs_dropped", 0)
    ))).run(conn)
    r.dbCreate(dbname).run(conn)
    conn.use(dbname)
    /* create status table */
    r.tableCreate(statusTableName).run(conn)
    /* save current status */
    r.table(statusTableName).insert(r.hashMap(statusFieldName, currentStatus)).run(conn)
    /* create other tables */
    Tables.tableNames
        .filterNot(statusTableName.equals)
        .foreach(t => unit(r.tableCreate(t).run(conn)))
    /* update status */
    r.table(statusTableName).update(r.hashMap(statusFieldName, nextStatus)).run(conn)
  }
}
